### PR TITLE
Fix stack overflow when generating debuginfo for 'recursive' type

### LINKED ISSUE #58463: compiler stack overflow compiling run-pass/issues/issue-29466.rs and run-pass/issues/issue-36792.rs with debuginfo
With `debuginfo-tests` set to true in config.toml, running `python x.py test src/test/run-pass` fails with:

```
---- [run-pass] run-pass/issues/issue-36792.rs stdout ----

error: test compilation failed although it shouldn't!
status: signal: 6
command: "/mnt/rust/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/mnt/rust/src/test/run-pass/issues/issue-36792.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-C" "prefer-dynamic" "-o" "/mnt/rust/build/x86_64-unknown-linux-gnu/test/run-pass/issues/issue-36792/a" "-Crpath" "-O" "-g" "-Zunstable-options" "-Lnative=/mnt/rust/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/mnt/rust/build/x86_64-unknown-linux-gnu/test/run-pass/issues/issue-36792/auxiliary"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------

thread 'rustc' has overflowed its stack
fatal runtime error: stack overflow

------------------------------------------

thread '[run-pass] run-pass/issues/issue-36792.rs' panicked at 'explicit panic', src/tools/compiletest/src/runtest.rs:3295:9
stack backtrace:
   0:     0x5572ef0bc323 - std::sys::unix::backtrace::tracing::imp::unwind_backtrace::he5ca7e8a91fc68be
                               at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:39
   1:     0x5572ef0b7df8 - std::sys_common::backtrace::_print::h60e068884b12931c
                               at src/libstd/sys_common/backtrace.rs:70
   2:     0x5572ef0bae32 - std::panicking::default_hook::{{closure}}::hc983185de2298168
                               at src/libstd/sys_common/backtrace.rs:58
                               at src/libstd/panicking.rs:200
   3:     0x5572ef0bab27 - std::panicking::default_hook::h6b97c91c835a5a90
                               at src/libstd/panicking.rs:209
   4:     0x5572ef0bb540 - std::panicking::rust_panic_with_hook::h96c73571b407939c
                               at src/libstd/panicking.rs:478
   5:     0x5572eefa4284 - std::panicking::begin_panic::hdcb9c5c9848385af
   6:     0x5572eef877d3 - compiletest::runtest::ProcRes::fatal::hd4564792a1dcc8c2
   7:     0x5572eef83072 - compiletest::runtest::TestCx::fatal_proc_rec::h8dada00ad6f542ef
   8:     0x5572eef71d3c - compiletest::runtest::TestCx::check_if_test_should_compile::hbff897c11854a61e
   9:     0x5572eef6ea67 - compiletest::runtest::TestCx::run_revision::he8f5a12ab03f9f49
  10:     0x5572eef67b87 - compiletest::runtest::run::hcf95bec718f18726
  11:     0x5572eefc30ba - <F as alloc::boxed::FnBox<A>>::call_box::h52e5773184df5b88
  12:     0x5572ef080012 - <F as alloc::boxed::FnBox<A>>::call_box::h9444f32827504df2
                               at src/libtest/lib.rs:1468
                               at /rustc/d1add97236b64048294692d91fe82b601577dd1f/src/liballoc/boxed.rs:734
  13:     0x5572ef0c4409 - __rust_maybe_catch_panic
                               at src/libpanic_unwind/lib.rs:92
  14:     0x5572ef096d37 - test::run_test::run_test_inner::{{closure}}::hb362655f70a2a56a
                               at /rustc/d1add97236b64048294692d91fe82b601577dd1f/src/libstd/panicking.rs:276
                               at /rustc/d1add97236b64048294692d91fe82b601577dd1f/src/libstd/panic.rs:388
                               at src/libtest/lib.rs:1430
  15:     0x5572ef079004 - std::sys_common::backtrace::__rust_begin_short_backtrace::hdfd30b3004ce04df
                               at /rustc/d1add97236b64048294692d91fe82b601577dd1f/src/libstd/sys_common/backtrace.rs:135
  16:     0x5572ef0795f4 - std::panicking::try::do_call::h1de50de762eb8517
                               at /rustc/d1add97236b64048294692d91fe82b601577dd1f/src/libstd/thread/mod.rs:469
                               at /rustc/d1add97236b64048294692d91fe82b601577dd1f/src/libstd/panic.rs:309
                               at /rustc/d1add97236b64048294692d91fe82b601577dd1f/src/libstd/panicking.rs:297
  17:     0x5572ef0c4409 - __rust_maybe_catch_panic
                               at src/libpanic_unwind/lib.rs:92
  18:     0x5572ef07fe1c - <F as alloc::boxed::FnBox<A>>::call_box::h49e936a4ce10b9e3
                               at /rustc/d1add97236b64048294692d91fe82b601577dd1f/src/libstd/panicking.rs:276
                               at /rustc/d1add97236b64048294692d91fe82b601577dd1f/src/libstd/panic.rs:388
                               at /rustc/d1add97236b64048294692d91fe82b601577dd1f/src/libstd/thread/mod.rs:468
                               at /rustc/d1add97236b64048294692d91fe82b601577dd1f/src/liballoc/boxed.rs:734
  19:     0x5572ef0c386d - std::sys::unix::thread::Thread::new::thread_start::h9819c9db0b318724
                               at /rustc/d1add97236b64048294692d91fe82b601577dd1f/src/liballoc/boxed.rs:744
                               at src/libstd/sys_common/thread.rs:14
                               at src/libstd/sys/unix/thread.rs:81
  20:     0x7f80b3d026da - start_thread
  21:     0x7f80b381388e - __clone
  22:                0x0 - <unknown>

---- [run-pass] run-pass/issues/issue-29466.rs stdout ----

error: test compilation failed although it shouldn't!
status: signal: 6
command: "/mnt/rust/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/mnt/rust/src/test/run-pass/issues/issue-29466.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-C" "prefer-dynamic" "-o" "/mnt/rust/build/x86_64-unknown-linux-gnu/test/run-pass/issues/issue-29466/a" "-Crpath" "-O" "-g" "-Zunstable-options" "-Lnative=/mnt/rust/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/mnt/rust/build/x86_64-unknown-linux-gnu/test/run-pass/issues/issue-29466/auxiliary"
stdout:
------------------------------------------

------------------------------------------
stderr:
------------------------------------------

thread '<unknown>' has overflowed its stack
fatal runtime error: stack overflow

------------------------------------------

thread '[run-pass] run-pass/issues/issue-29466.rs' panicked at 'explicit panic', src/tools/compiletest/src/runtest.rs:3295:9
stack backtrace:
   0:     0x5572ef0bc323 - std::sys::unix::backtrace::tracing::imp::unwind_backtrace::he5ca7e8a91fc68be
                               at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:39
   1:     0x5572ef0b7df8 - std::sys_common::backtrace::_print::h60e068884b12931c
                               at src/libstd/sys_common/backtrace.rs:70
   2:     0x5572ef0bae32 - std::panicking::default_hook::{{closure}}::hc983185de2298168
                               at src/libstd/sys_common/backtrace.rs:58
                               at src/libstd/panicking.rs:200
   3:     0x5572ef0bab27 - std::panicking::default_hook::h6b97c91c835a5a90
                               at src/libstd/panicking.rs:209
   4:     0x5572ef0bb540 - std::panicking::rust_panic_with_hook::h96c73571b407939c
                               at src/libstd/panicking.rs:478
   5:     0x5572eefa4284 - std::panicking::begin_panic::hdcb9c5c9848385af
   6:     0x5572eef877d3 - compiletest::runtest::ProcRes::fatal::hd4564792a1dcc8c2
   7:     0x5572eef83072 - compiletest::runtest::TestCx::fatal_proc_rec::h8dada00ad6f542ef
   8:     0x5572eef71d3c - compiletest::runtest::TestCx::check_if_test_should_compile::hbff897c11854a61e
   9:     0x5572eef6ea67 - compiletest::runtest::TestCx::run_revision::he8f5a12ab03f9f49
  10:     0x5572eef67b87 - compiletest::runtest::run::hcf95bec718f18726
  11:     0x5572eefc30ba - <F as alloc::boxed::FnBox<A>>::call_box::h52e5773184df5b88
  12:     0x5572ef080012 - <F as alloc::boxed::FnBox<A>>::call_box::h9444f32827504df2
                               at src/libtest/lib.rs:1468
                               at /rustc/d1add97236b64048294692d91fe82b601577dd1f/src/liballoc/boxed.rs:734
  13:     0x5572ef0c4409 - __rust_maybe_catch_panic
                               at src/libpanic_unwind/lib.rs:92
  14:     0x5572ef096d37 - test::run_test::run_test_inner::{{closure}}::hb362655f70a2a56a
                               at /rustc/d1add97236b64048294692d91fe82b601577dd1f/src/libstd/panicking.rs:276
                               at /rustc/d1add97236b64048294692d91fe82b601577dd1f/src/libstd/panic.rs:388
                               at src/libtest/lib.rs:1430
  15:     0x5572ef079004 - std::sys_common::backtrace::__rust_begin_short_backtrace::hdfd30b3004ce04df
                               at /rustc/d1add97236b64048294692d91fe82b601577dd1f/src/libstd/sys_common/backtrace.rs:135
  16:     0x5572ef0795f4 - std::panicking::try::do_call::h1de50de762eb8517
                               at /rustc/d1add97236b64048294692d91fe82b601577dd1f/src/libstd/thread/mod.rs:469
                               at /rustc/d1add97236b64048294692d91fe82b601577dd1f/src/libstd/panic.rs:309
                               at /rustc/d1add97236b64048294692d91fe82b601577dd1f/src/libstd/panicking.rs:297
  17:     0x5572ef0c4409 - __rust_maybe_catch_panic
                               at src/libpanic_unwind/lib.rs:92
  18:     0x5572ef07fe1c - <F as alloc::boxed::FnBox<A>>::call_box::h49e936a4ce10b9e3
                               at /rustc/d1add97236b64048294692d91fe82b601577dd1f/src/libstd/panicking.rs:276
                               at /rustc/d1add97236b64048294692d91fe82b601577dd1f/src/libstd/panic.rs:388
                               at /rustc/d1add97236b64048294692d91fe82b601577dd1f/src/libstd/thread/mod.rs:468
                               at /rustc/d1add97236b64048294692d91fe82b601577dd1f/src/liballoc/boxed.rs:734
  19:     0x5572ef0c386d - std::sys::unix::thread::Thread::new::thread_start::h9819c9db0b318724
                               at /rustc/d1add97236b64048294692d91fe82b601577dd1f/src/liballoc/boxed.rs:744
                               at src/libstd/sys_common/thread.rs:14
                               at src/libstd/sys/unix/thread.rs:81
  20:     0x7f80b3d026da - start_thread
  21:     0x7f80b381388e - __clone
  22:                0x0 - <unknown>


failures:
    [run-pass] run-pass/issues/issue-29466.rs
    [run-pass] run-pass/issues/issue-36792.rs
```

### PR DESCRIPTION
**Pull request by Aaron1011**
By using 'impl trait', it's possible to create a self-referential
type as follows:

fn foo() -> impl Copy { foo }

This is a function which returns itself.
Normally, the signature of this function would be impossible
to write - it would look like 'fn foo() -> fn() -> fn() ...'
e.g. a function which returns a function, which returns a function...

Using 'impl trait' allows us to avoid writing this infinitely long
type. While it's useless for practical purposes, it does compile and run

However, issues arise when we try to generate llvm debuginfo for such a
type. All 'impl trait' types (e.g. ty::Opaque) are resolved when we
generate debuginfo, which can lead to us recursing back to the original
'fn' type when we try to process its return type.

To resolve this, I've modified debuginfo generation to account for these
kinds of weird types. Unfortunately, there's no 'correct' debuginfo that
we can generate - 'impl trait' does not exist in debuginfo, and this
kind of recursive type is impossible to directly represent.

To ensure that we emit *something*, this commit emits dummy
debuginfo/type names whenever it encounters a self-reference. In
practice, this should never happen - it's just to ensure that we can
emit some kind of debuginfo, even if it's not particularly meaningful

Fixes #58463

### CODE REVIEW

File: src/librustc_codegen_llvm/debuginfo/metadata.rs [hunk 1]
```diff
@@ -596,17 +622,51 @@ pub fn type_metadata(
...
+                    // type is going to see *somthing* weird - the only
+                    // question is what exactly it will see
+                    let (size, align) = cx.size_and_align_of(t);
+                    llvm::LLVMRustDIBuilderCreateBasicType(
+                        DIB(cx),
+                        SmallCStr::new("<recur_type>").as_ptr(),
+                        size.bits(),
+                        align.bits() as u32,
+                        DW_ATE_unsigned)
+
```
Comment (oli-obk) [inline src/librustc_codegen_llvm/debuginfo/metadata.rs]: there are a bunch of empty lines here

File: src/librustc_codegen_llvm/debuginfo/metadata.rs [hunk 2]
```diff
@@ -116,6 +116,32 @@ impl TypeMap<'ll, 'tcx> {
...
+    // via LLVM debuginfo. As a workaround,
+    // we register a temporary Ty to metadata mapping
+    // for the function before we compute its actual metadat.a
+    // If the metadata computation ends up recursing back to the
+    // original function, it will use the temporary mapping
+    // for the inner self-reference, preventing us from
+    // recursing forever.
+    //
+    // This function is used to remove the temporary metadata
+    // mapping after we've computed the actual metadat
```
Comment (oli-obk) [inline src/librustc_codegen_llvm/debuginfo/metadata.rs]:
```suggestion
    // mapping after we've computed the actual metadata
```

File: src/librustc_codegen_llvm/debuginfo/metadata.rs [hunk 3]
```diff
@@ -596,17 +622,51 @@ pub fn type_metadata(
...
+
+            let type_map = &debug_context(cx).type_map;
+            type_map.borrow_mut().register_type_with_metadata(t, temp_type);
+
+            let fn_metadata = subroutine_type_metadata(cx,
+                                                       unique_type_id,
+                                                       t.fn_sig(cx.tcx),
+                                                       usage_site_span).metadata;
+
+            type_map.borrow_mut().remove_type(t);
```
Comment (oli-obk) [inline src/librustc_codegen_llvm/debuginfo/metadata.rs]: why are we registering this only temporarily and don't just leave it in?

Comment (Aaron1011) [reply src/librustc_codegen_llvm/debuginfo/metadata.rs]: The temporary metadata is only used in the case where we recourse back to processing the same type.

The actual type will be inserted into the map at the bottom of the function.

**Comment by Aaron1011** [comment]
@oli-obk I've made the changes you requested

**Comment by Centril** [comment]
Speculatively assigning blame of failure in https://github.com/rust-lang/rust/pull/59538#issuecomment-478198475, @bors r-

**Comment by Aaron1011** [comment]
@Centril: Fixed

File: src/librustc_codegen_llvm/debuginfo/type_names.rs [hunk 4]
```diff
@@ -179,6 +180,18 @@ pub fn push_debuginfo_type_name<'a, 'tcx>(cx: &CodegenCx<'a, 'tcx>,
...
+
+
+            // We only keep the type in 'visited'
+            // for the duration of the body of this method.
+            // It's fine for a particular function type
+            // to show up multiple times in one overall type
+            // (e.g. MyType<fn() -> u8, fn() -> u8>
+            //
+            // We only care about avoiding recursing
+            // directly back to the type we're currentlu
```
Comment (pnkfelix) [inline src/librustc_codegen_llvm/debuginfo/type_names.rs]: typo: "currentlu" should be "currently"

**Comment by Aaron1011** [comment]
@pnkfelix Fixed

**Comment by oli-obk** [comment]
Static items can also be recursive via `impl Trait`, but currently fail to compile due to other reasons https://github.com/rust-lang/rust/issues/59631

This PR's changes can be extended to cover statics once statics actually hit the problematic situation and don't error out early.

@bors r+

**Pull request merged**